File: 234-palindrome-linked-list/234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        
        string s;
        
        ListNode* temp = head;
        
        while(temp!=NULL){
            s+=temp->val;
            temp = temp->next;
        }
        
        int start  = 0;
        int end = s.size()-1;
        
        while(start<=end){
            if(s[start++]!=s[end--]) return false;
        }
        
        return true;
        
        
    }
};

File: 234-palindrome-linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def reverse(self,head):
        
        curr = head
        prev = None
        
        while(curr):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        
        return prev
    
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        
        if head is None or head.next is None:
            return True
        
        fast = head
        slow = head
        
        while(fast.next and fast.next.next):
            slow = slow.next
            fast = fast.next.next
        
        #slow at end of 1st half
        # reverse 2nd half
        
        second_start = self.reverse(slow.next)
        
        p1 = head
        p2 = second_start
        
        ans = True
        
        while(ans and p2!=None):
            if(p1.val!=p2.val):
                return False
            p1 = p1.next
            p2 = p2.next
        
        
        # slow.next = self.reverse(second_start)
        
        return True
        
        
        
        
        